staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go

```diff
@@ -213,8 +270,39 @@ func (s *Serializer) Decode(originalData []byte, gvk *schema.GroupVersionKind, i
 		return nil, actual, err
 	}
 
-	if err := caseSensitiveJsonIterator.Unmarshal(data, obj); err != nil {
-		return nil, actual, err
+	// If the provided decoding mode is strict, first check if there was a duplicate error earlier in the original JSON
+	// or YAML data, and in that case, return it as a DuplicateFieldError. If that's not a problem, decode using the strict
+	// json-iter configuration, and return a possible error as UnknownFieldError.
+	// If non-strict mode just use the "normal" non-strict json-iter decoder.
+	if s.strict {
+		if yamlToJSONError != nil {
+			str := yamlToJSONError.Error()
+			// Detect DuplicateFieldError, otherwise return the generic StrictDecoderError.
+			// In the future we might want to handle this in a better way. Currently the used libraries
+			// do not return typed errors.
+			if strings.Contains(str, `already set in map`) {
```

Comment:
let's not take dependencies on error message strings. 

if we don't care about efficiency for the strict path yet (and I don't think we do until we are able to do a single decode pass and get back duplicate/unknown field info independent of completing the decoding), would it be simpler to do the following:

```go
// make a copy of the object we're going to unmarshal into
strictObj := obj.DeepCopy()

... existing decode logic, return error if fails

// do additional checks
if strict {
  // catch duplicate fields
  strictJSONData, err := yaml.YAMLToJSONStrict(originalData)
  if err != nil {
    return runtime.NewStrictDecoderError(err.Error(), *actual, originalData)
  }
  // catch unknown fields
  err = caseSensitiveStrictJsonIterator.Unmarshal(strictJSONData, strictObj)
  if err != nil {
    return runtime.NewStrictDecoderError(err.Error(), *actual, originalData)
  }
  // maybe catch differences in decoding when strict? not sure about this...
  if !reflect.DeepEqual(strictObj, obj) {
    return runtime.NewStrictDecoderError(strict decoding diff?, *actual, originalData)
  }
}
```

Reply:
that would also help ensure that a strict serializer never returns a different object than the normal decoder path in a success case